quantization/src/encoded_vectors_binary.rs

```diff
@@ -5,81 +5,127 @@ use crate::{
     VectorParameters,
 };
 use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
 use std::path::Path;
 
-type BitsStoreType = u128;
-
-const BITS_STORE_TYPE_SIZE: usize = std::mem::size_of::<BitsStoreType>() * 8;
-
-pub struct EncodedVectorsBin<TStorage: EncodedStorage> {
+pub struct EncodedVectorsBin<TBitsStoreType: BitsStoreType, TStorage: EncodedStorage> {
     encoded_vectors: TStorage,
     metadata: Metadata,
+    bits_store_type: PhantomData<TBitsStoreType>,
 }
 
-pub struct EncodedBinVector {
-    encoded_vector: Vec<BitsStoreType>,
+pub struct EncodedBinVector<TBitsStoreType: BitsStoreType> {
+    encoded_vector: Vec<TBitsStoreType>,
 }
 
 #[derive(Serialize, Deserialize)]
 struct Metadata {
     vector_parameters: VectorParameters,
 }
 
-impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
-    pub fn encode<'a>(
-        orig_data: impl Iterator<Item = impl AsRef<[f32]> + 'a> + Clone,
-        mut storage_builder: impl EncodedStorageBuilder<TStorage>,
-        vector_parameters: &VectorParameters,
-        stop_condition: impl Fn() -> bool,
-    ) -> Result<Self, EncodingError> {
-        debug_assert!(validate_vector_parameters(orig_data.clone(), vector_parameters).is_ok());
+pub trait BitsStoreType:
```

Comment:
To separate dense and multivector behaviour, I defined here a generic type which is `u128` for dense and `u8` for multivector